### PR TITLE
[TLS] Timelock Stability: Single CachedThreadPool for LockServiceImpls

### PR DESCRIPTION
**Pull request by gmaretic**
**Goals (and why)**:
Achieve better thread reuse and reduce overhead of creating an executor service per client on gaining leadership

**Implementation Description (bullets)**:
Pretty straightforward

**Testing (What was existing testing like?  What have you done to improve it?)**:
Smoke tested

**Concerns (what feedback would you like?)**:
Did not address the single threaded executors for async lock, as that is more tricky to do and seems less beneficial.

### CODE REVIEW

File: lock-impl/src/main/java/com/palantir/lock/impl/LockServiceImpl.java [hunk 1]
```diff
@@ -1060,16 +1068,16 @@ private StringBuilder getGeneralLockStats() {
...
-            callOnClose.run();
+            shutDownTask.run();
         }
     }
 
-    private void wakeIndefiniteBlockers() {
-        for (Thread blocked : indefinitelyBlockingThreads) {
-            blocked.interrupt();
+    private void shutDown(InjectableResource<ExecutorService> executor, Runnable callOnClose) {
+        if (!executor.isInjected()) {
```
Comment (felixdesouza) [inline lock-impl/src/main/java/com/palantir/lock/impl/LockServiceImpl.java]: with the `Owned` suggestion, it'll look like:

```java
private void shutDown() {
    if (executor.isOwned()) {
        executor.resource().shutdownNow();
    }
    indefinitelyBlockingThreads.forEach(Thread::interrupt);
	callOnClose.run();
}
```

with the `executor` and `callOnClose` remaining as fields

Comment (felixdesouza) [reply lock-impl/src/main/java/com/palantir/lock/impl/LockServiceImpl.java]: or if wrapped:

```java
private void shutDown() {
  executor.shutdownNow();
  indefinitelyBlockingThreads.forEach(Thread::interrupt);
  callOnClose.run();
}
```

File: atlasdb-commons/src/main/java/com/palantir/util/InjectableResource.java [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.util;
+
+import org.immutables.value.Value;
+
+@Value.Immutable
+public interface InjectableResource<R> {
```
[the remark this answers is not in the document]
Comment (felixdesouza) [reply atlasdb-commons/src/main/java/com/palantir/util/InjectableResource.java]: Personally, I'm in favour of the latter.

File: timelock-agent/src/main/java/com/palantir/timelock/paxos/LockCreator.java [hunk 3]
```diff
@@ -29,6 +32,8 @@
 public class LockCreator {
     private final Supplier<TimeLockRuntimeConfiguration> runtime;
     private final TimeLockDeprecatedConfiguration deprecated;
+    private final ExecutorService sharedExecutor = PTExecutors
```
[the remark this answers is not in the document]
Comment (gmaretic) [reply timelock-agent/src/main/java/com/palantir/timelock/paxos/LockCreator.java]: Yup, we should do this eventually, but in a separate PR

**Comment by gmaretic** [comment]
@felixdesouza good to merge?

File: atlasdb-commons/src/main/java/com/palantir/util/Owned.java [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+package com.palantir.util;
+
+import org.immutables.value.Value;
+
+@Value.Immutable
+public interface Owned<R> {
+    R resource();
+    boolean isOwned();
+
+    static <T> Owned<T> owned(T ownedResource) {
```
Comment (felixdesouza) [inline atlasdb-commons/src/main/java/com/palantir/util/Owned.java]: nit: `T` -> `R`

**Pull request closed without merging**